File: insight-brain-frontend/src/main/frontend/applicationReport/applicationReportFilter.js
/*
 * Copyright (c) 2011-present Sonatype, Inc. All rights reserved.
 * Includes the third-party code listed at http://links.sonatype.com/products/clm/attributions.
 * "Sonatype" is a trademark of Sonatype, Inc.
 */
import template from './applicationReportFilter.html';
import { policyTypes } from '../dashboard/filter/staticFilterEntries';
import { equals, head, last, map, pick, range, reduce, reject } from 'ramda';
import { fetchStageTypes } from '../stages/stagesActions';
import { lookup, setToArray, union } from '../util/jsUtil';
import { faArrowToRight } from '@fortawesome/pro-solid-svg-icons';

export default {
  template: template,
  controllerAs: 'vm',
  controller: ApplicationReportFilterController,
};

// Map from checkbox option id to violationState filter set
const violationStateCheckboxFilterMapping = {
  notViolating: new Set(['notViolating']),
  open: new Set(['open']),
  waived: new Set(['waived', 'waived+grandfathered']),
  grandfathered: new Set(['grandfathered', 'waived+grandfathered']),
};

export function ApplicationReportFilterController(
  $scope,
  $ngRedux,
  applicationReportActions
) {
  const vm = this;

  Object.assign(vm, {
    availableProprietaryFilterOptions: [
      { id: false, name: 'Non-Proprietary' },
      { id: true, name: 'Proprietary' },
    ],

    availableMatchStateFilterOptions: [
      { id: 'exact', name: 'Exact' },
      { id: 'similar', name: 'Similar' },
      { id: 'unknown', name: 'Unknown' },
    ],

    availableViolationStateFilterOptions: [
      { id: 'notViolating', name: 'Not Violating' },
      { id: 'open', name: 'Open' },
      { id: 'waived', name: 'Waived' },
      { id: 'grandfathered', name: 'Grandfathered' },
    ],

    availableDependencyTypeFilterOptions: [
      { id: 'direct', name: 'Direct Dependencies' },
      { id: 'transitive', name: 'Transitive Dependencies' },
      { id: 'unknown', name: 'Unknown' },
    ],

    availablePolicyTypeFilterOptions: policyTypes,

    violationStateCheckedIds: new Set(),

    policyThreatLevelFilterSelectedRange: undefined,

    faArrowToRightIcon: faArrowToRight,

    $onInit() {
      const actions = {
        ...pick(
          [
            'setAggregateReportEntries',
            'setExactValueFilter',
            'reevaluateReport',
            'loadInnerSourceReports',
            'toggleFilterSidebar',
          ],
          applicationReportActions
        ),
        fetchStageTypes,
      };

      vm.unsubscribe = $ngRedux.connect(mapStateToThis, actions)(vm);
      vm.fetchStageTypes('action');
      document.addEventListener('keydown', vm.escapeKeyHandler);
      document.addEventListener('mousedown', vm.offClickHandler);

      $scope.$watch(
        'vm.exactValueFilters.derivedViolationState',
        function (derivedViolationState) {
          const violationStateFilter = derivedViolationState || new Set(),
            // the 'waived+grandfathered' value is redundant for these purposes, and the other possible values
            // all map perfectly to the checkbox ids
            checkedIds = reject(
              equals('waived+grandfathered'),
              setToArray(violationStateFilter)
            );

          vm.violationStateCheckedIds = new Set(checkedIds);
        }
      );

      $scope.$watch(
        'vm.exactValueFilters.policyThreatLevel',
        function (allowedValues) {
          vm.policyThreatLevelFilterSelectedRange = toSelectedRange(
            allowedValues
          );
        }
      );
    },

    $onDestroy() {
      vm.unsubscribe();
      document.removeEventListener('keydown', vm.escapeKeyHandler);
      document.removeEventListener('mousedown', vm.offClickHandler);
    },

    setProprietaryFilterOptions(selectedIds) {
      vm.setExactValueFilter('proprietary', selectedIds);
    },

    setMatchStateFilterOptions(selectedIds) {
      vm.setExactValueFilter('matchState', selectedIds);
    },

    setViolationStateFilterOptions(selectedIds) {
      const selectedFilters = map(
          lookup(violationStateCheckboxFilterMapping),
          setToArray(selectedIds)
        ),
        mergedFilter = reduce(union, new Set(), selectedFilters);

      vm.setExactValueFilter('derivedViolationState', mergedFilter);
    },

    setPolicyTypeFilterOptions(selectedIds) {
      vm.setExactValueFilter('policyThreatCategory', selectedIds);
    },

    setDependencyTypeFilterOptions(selectedIds) {
      vm.setExactValueFilter('derivedDependencyType', selectedIds);
    },

    setPolicyThreatLevelFilter(selectedRange) {
      vm.setExactValueFilter(
        'policyThreatLevel',
        fromSelectedRange(selectedRange)
      );
    },

    closeSideBarFilterIfOpen() {
      if (vm.filterSidebarOpen) {
        vm.toggleFilterSidebar(false);
      }
    },

    escapeKeyHandler({ key }) {
      if (key === 'Escape' || key === 'Esc') {
        vm.closeSideBarFilterIfOpen();
      }
    },

    offClickHandler(event) {
      if (!vm.filterSidebarOpen) {
        return;
      }

      const filterMainElement = document.getElementById(
        'application-report-sidebar'
      );
      const isClickOnReportFilter = filterMainElement.contains(event.target);
      if (!isClickOnReportFilter) {
        vm.closeSideBarFilterIfOpen();
      }
    },
  });
}

function mapStateToThis(state) {
  return pick(
    [
      'policyTypeFilterEnabled',
      'aggregate',
      'exactValueFilters',
      'reportParameters',
      'filterSidebarOpen',
    ],
    state.applicationReport || {}
  );
}

ApplicationReportFilterController.$inject = [
  '$scope',
  '$ngRedux',
  'applicationReportActions',
];

function toSelectedRange(allowedValues) {
  if (allowedValues && allowedValues.size) {
    const rangeArray = setToArray(allowedValues);
    return [Math.min(...rangeArray), Math.max(...rangeArray)];
  }
  // if filter is empty - set slider to full range
  return [0, 10];
}

function fromSelectedRange(selectedRange) {
  // if whole range is selected - don't do any filtering
  return equals([0, 10], selectedRange)
    ? new Set()
    : new Set(range(head(selectedRange), last(selectedRange) + 1));
}

File: insight-brain-java-functional-test/src/main/java/com/sonatype/clm/testing/functional/elements/ApplicationReportFilter.java
/*
 * Copyright (c) 2011-present Sonatype, Inc. All rights reserved.
 * Includes the third-party code listed at http://links.sonatype.com/products/clm/attributions.
 * "Sonatype" is a trademark of Sonatype, Inc.
 */
package com.sonatype.clm.testing.functional.elements;

import com.sonatype.clm.testing.functional.BasicElement;

import com.codeborne.selenide.SelenideElement;

public class ApplicationReportFilter
    extends BasicElement<ApplicationReportFilter>
{
  public ApplicationReportFilter() {
    super("#iq-component-filter-popover");
  }

  public SelenideElement closeButton() {
    return child(".iq-popover-header__close-btn");
  }

  public ProprietaryFilter proprietaryFilter() {
    return new ProprietaryFilter(childSelector("#proprietary-filter"));
  }

  public InnerSourceFilter innerSourceFilter() {
    return new InnerSourceFilter(childSelector("#inner-source-filter"));
  }

  public MatchStateFilter matchStateFilter() {
    return new MatchStateFilter(childSelector("#match-state-filter"));
  }

  public ViolationStateFilter violationStateFilter() {
    return new ViolationStateFilter(childSelector("#violation-state-filter"));
  }

  public DependencyTypeFilter dependencyTypeFilter() {
    return new DependencyTypeFilter(childSelector("#dependency-type-filter"));
  }

  public PolicyTypeFilter policyTypeFilter() {
    return new PolicyTypeFilter(childSelector("#policy-type-filter"));
  }

  public static class ProprietaryFilter
      extends NxTreeViewMultiSelect
  {
    public ProprietaryFilter(String selector) {
      super(selector);
    }

    public NxCheckbox nonProprietary() {
      return super.checkboxItem(2);
    }

    public NxCheckbox proprietary() {
      return super.checkboxItem(3);
    }
  }

  public static class InnerSourceFilter
      extends NxTreeViewMultiSelect
  {
    public InnerSourceFilter(String selector) {
      super(selector);
    }

    public NxCheckbox nonInnerSource() {
      return super.checkboxItem(2);
    }

    public NxCheckbox innerSource() {
      return super.checkboxItem(3);
    }
  }

  public static class MatchStateFilter
      extends NxTreeViewMultiSelect
  {
    public MatchStateFilter(String selector) {
      super(selector);
    }

    public NxCheckbox exact() {
      return super.checkboxItem(2);
    }

    public NxCheckbox similar() {
      return super.checkboxItem(3);
    }

    public NxCheckbox unknown() {
      return super.checkboxItem(4);
    }
  }

  public static class ViolationStateFilter
      extends NxTreeViewMultiSelect
  {
    public ViolationStateFilter(String selector) {
      super(selector);
    }

    public NxCheckbox notViolating() {
      return super.checkboxItem(2);
    }

    public NxCheckbox open() {
      return super.checkboxItem(3);
    }

    public NxCheckbox waived() {
      return super.checkboxItem(4);
    }

    public NxCheckbox grandfathered() {
      return super.checkboxItem(5);
    }
  }

  public static class DependencyTypeFilter
      extends NxTreeViewMultiSelect
  {
    public DependencyTypeFilter(final String selector) {
      super(selector);
    }

    public NxCheckbox direct() {
      return super.checkboxItem(2);
    }

    public NxCheckbox transitive() {
      return super.checkboxItem(3);
    }

    public NxCheckbox unknown() {
      return super.checkboxItem(4);
    }
  }

  public static class PolicyTypeFilter
      extends NxTreeViewMultiSelect
  {
    public PolicyTypeFilter(String selector) {
      super(selector);
    }

    public NxCheckbox security() {
      return super.checkboxItem(2);
    }

    public NxCheckbox license() {
      return super.checkboxItem(3);
    }

    public NxCheckbox quality() {
      return super.checkboxItem(4);
    }

    public NxCheckbox other() {
      return super.checkboxItem(5);
    }
  }
}
